### PR TITLE
[flutter_tool] Prevent accidental calls to io.exit in unit tests

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

To prevent a scenario such as #46142 for occurring again in the future, prevent calls to dart:io's exit when asserts are enabled within a unit test. This is detected by invoking `expect` and checking if a StateError is thrown

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/io.dart [hunk 1]
```diff
@@ -92,12 +93,34 @@ ExitFunction _exitFunction = _defaultExitFunction;
 
 /// Exits the process.
 ///
+/// Throws [StateError] if assertions are enabled and the dart:io exit
+/// is still active when called. This may indicate exit was called in
+/// a test without being configured correctly. This behavior can be
+/// removed by setting the `FLUTTER_TEST` environment
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/io.dart]: Looks like the comment needs to be updated.

File: packages/flutter_tools/lib/src/base/io.dart [hunk 2]
```diff
@@ -92,12 +93,34 @@ ExitFunction _exitFunction = _defaultExitFunction;
...
 /// This is analogous to the `exit` function in `dart:io`, except that this
 /// function may be set to a testing-friendly value by calling
 /// [setExitFunctionForTests] (and then restored to its default implementation
 /// with [restoreExitFunction]). The default implementation delegates to
 /// `dart:io`.
-ExitFunction get exit => _exitFunction;
+ExitFunction get exit {
+  assert(() {
+    if (_exitFunction == io.exit) {
+      try {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/io.dart]: Consider pulling the `expect()` check into a helper method, and simplifying here as in the last PR.

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/io.dart]: Done

File: packages/flutter_tools/lib/src/base/io.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/io.dart]: Done

File: packages/flutter_tools/lib/src/base/io.dart [hunk 3]
```diff
@@ -92,12 +93,33 @@ ExitFunction _exitFunction = _defaultExitFunction;
...
 /// This is analogous to the `exit` function in `dart:io`, except that this
 /// function may be set to a testing-friendly value by calling
 /// [setExitFunctionForTests] (and then restored to its default implementation
 /// with [restoreExitFunction]). The default implementation delegates to
 /// `dart:io`.
-ExitFunction get exit => _exitFunction;
+ExitFunction get exit {
+  assert(
+    _exitFunction != io.exit || !_inUnitTest(),
+    'io.exit was called with assertions active in a unit test'
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/io.dart]: nit: Does assert allow a comma after the message string?

**Review by zanderso — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/base/io.dart (hunk 3 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/base/io.dart]: Lets find out...

**Pull request merged**